oak_crypto/src/noise_nk/mod.rs

```diff
@@ -0,0 +1,267 @@
+        if *nonce > MAX_SEQUENCE {
+            return Err(Error::DecryptFailed);
+        }
+        let mut ret = [0u8; NONCE_LEN];
+        ret[NONCE_LEN - 4..].copy_from_slice(nonce.to_be_bytes().as_slice());
+        *nonce += 1;
+        Ok(ret)
+    }
+
+    pub fn encrypt(&mut self, plaintext: &[u8]) -> Result<Vec<u8>, Error> {
+        const PADDING_GRANULARITY: usize = 32;
+        static_assertions::const_assert!(PADDING_GRANULARITY < 256);
+        static_assertions::const_assert!((PADDING_GRANULARITY & (PADDING_GRANULARITY - 1)) == 0);
+
+        let mut padded_size: usize = plaintext.len();
+        if padded_size > (1usize << 28) {
+            return Err(Error::DataTooLarge(padded_size));
+        }
+        padded_size += 1; // padding-length byte
+        padded_size = (padded_size + PADDING_GRANULARITY - 1) & !(PADDING_GRANULARITY - 1);
+
+        let mut padded_encrypt_data = Vec::with_capacity(padded_size);
+        padded_encrypt_data.extend_from_slice(plaintext);
+        padded_encrypt_data.resize(padded_size, 0u8);
+        let num_zeros = padded_size - plaintext.len() - 1;
+        padded_encrypt_data[padded_size - 1] = num_zeros as u8;
+
+        rustcrypto::aes_256_gcm_seal_in_place(
+            &self.write_key,
+            &Self::next_nonce(&mut self.write_nonce)?,
+            &[],
+            &mut padded_encrypt_data,
+        );
+        Ok(padded_encrypt_data)
+    }
+
+    pub fn decrypt(&mut self, ciphertext: &[u8]) -> Result<Vec<u8>, Error> {
+        let plaintext = rustcrypto::aes_256_gcm_open_in_place(
+            &self.read_key,
+            &Self::next_nonce(&mut self.read_nonce)?,
+            &[],
+            Vec::from(ciphertext),
+        )
+        .map_err(|_| Error::DecryptFailed)?;
+
+        // Plaintext must have a padding byte, and the unpadded length must be
+        // at least one.
+        if plaintext.is_empty() || (plaintext[plaintext.len() - 1] as usize) >= plaintext.len() {
+            return Err(Error::DecryptionPadding);
+        }
+        let unpadded_length = plaintext.len() - (plaintext[plaintext.len() - 1] as usize);
+        Ok(Vec::from(&plaintext[0..unpadded_length - 1]))
+    }
+}
+
+pub struct Response {
+    pub crypter: Crypter,
+    pub handshake_hash: [u8; 32],
+    pub response: Vec<u8>,
+}
+
+/// Performs the Responder side of the Noise protocol with the NK pattern.
+/// |identity_private_key_bytes| contains the private key scalar for the
+/// service's provisioned identity. |in_data| is provided by the Initiator and
+/// contains its ephemeral public key and encrypted payload.
+///
+/// The identity public key is computed from the private key, but could
+/// alternatively be stored separately to reduce computation if needed to
+/// reduce per-transaction computation.
+/// See https://noiseexplorer.com/patterns/NK/
+pub fn respond(identity_private_key_bytes: &[u8], in_data: &[u8]) -> Result<Response, Error> {
```

Comment:
The format of `in_data` does not seem obvious -- should it be documented somewhere? And perhaps create another function to assemble and split that into a struct?